crates/ra_hir_def/src/body/lower.rs

```diff
@@ -160,11 +162,28 @@ impl ExprCollector<'_> {
         id
     }
 
+    fn update_parent_map(
+        &mut self,
+        (parent_expr, children_exprs): (ExprId, Vec<ExprId>),
+    ) -> ExprId {
+        for child_expr in children_exprs {
+            self.body.parent_map.insert(child_expr, parent_expr);
+        }
+
+        parent_expr
+    }
+
     fn collect_expr(&mut self, expr: ast::Expr) -> ExprId {
+        let parent_and_children = self.collect_expr_inner(expr);
+        self.update_parent_map(parent_and_children)
+    }
+
+    fn collect_expr_inner(&mut self, expr: ast::Expr) -> (ExprId, Vec<ExprId>) {
```

Comment:
I'm not sure I follow. The parents aren't constructed until after the children have already been constructed. So as your descending, you don't have the parent id. If we're just trying to avoid the vec allocation, I can do the mapping inline.

Reply:
Let me know what you think about the macro version. I think this avoids adding anything extra to `ExprCollector` and also avoids allocing too many extra `Vec`s to pass back. I couldn't avoid cloning existing vecs though. Since I can't get the parent expr id until I've already passed the needed children expr id vecs to the constructing call.